File: packages/server/src/routes/content/content.ts
import { Router } from "express";

import isAuth from "../../middlewares/isAuth";
import { prisma } from "../../utils/prisma";
import { withPagination } from "../../utils/withPagination";
import { zMiddleware, zParse } from "../../utils/zParse";
import { jwtType } from "../auth/utils";
import {
  createContentSchema,
  deleteContentSchema,
  getContentListSchema,
  getContentSchema,
  updateContentSchema,
} from "./content.schema";
import field from "./field";

const content = Router();
// nested routes
content.use("/field", field);
// current level routes
content.get(
  "/",
  isAuth(jwtType.ACCESS),
  zMiddleware(getContentListSchema),
  async (req, res) => {
    const {
      query: { page, pageSize },
      body: { contentModelId },
      user: { id: userId },
    } = await zParse(getContentListSchema, req);

    const totalContent = await prisma.content.count();
    const content = await prisma.content.findMany({
      where: { contentModelId, userId },
      include: { FIeld: true },
      ...withPagination(page, pageSize),
    });

    return res
      .status(200)
      .json({ content, pagination: { page, pageSize, total: totalContent } });
  }
);

content.get(
  "/:id",
  isAuth(jwtType.ACCESS),
  zMiddleware(getContentSchema),
  async (req, res) => {
    const {
      params: { id },
      body: { contentModelId },
      user: { id: userId },
    } = await zParse(getContentSchema, req);
    const content = await prisma.content.findMany({
      where: { id, contentModelId, userId },
    });

    return res.status(200).json({ content });
  }
);

content.post(
  "/",
  isAuth(jwtType.ACCESS),
  zMiddleware(createContentSchema),
  async (req, res) => {
    const {
      body: { title, contentModelId },
      user: { id: userId },
    } = await zParse(createContentSchema, req);

    const content = await prisma.content.create({
      data: {
        title,
        contentModelId,
        userId,
      },
    });

    const fieldTypes = await prisma.fieldType.findMany({
      where: { id: contentModelId },
    });
    const { id: contentId } = content;
    for (const { id: fieldTypeId, title } of fieldTypes) {
      await prisma.fIeld.create({
        data: { fieldTypeId, contentId, title },
      });
    }

    return res.status(201).json({ content });
  }
);

content.put(
  "/:id",
  isAuth(jwtType.ACCESS),
  zMiddleware(updateContentSchema),
  async (req, res) => {
    const {
      body: { title, contentModelId },
      params: { id },
      user,
    } = await zParse(updateContentSchema, req);

    const updateContent = await prisma.content.update({
      where: {
        id,
        userId: user.id,
        contentModelId,
      },
      data: {
        title: title,
      },
    });
    return res.status(201).json({ content: updateContent });
  }
);

content.delete(
  "/:id",
  isAuth(jwtType.ACCESS),
  zMiddleware(deleteContentSchema),
  async (req, res) => {
    const {
      body: { contentModelId },
      params: { id },
      user: { id: userId },
    } = await zParse(deleteContentSchema, req);

    await prisma.content.delete({
      where: { id, contentModelId, userId },
    });

    return res.status(204).send();
  }
);

export default content;

File: apps/server/routes/content/content.go
package content

import (
	"fmt"
	"strconv"
	"svetozar12/headless-cms-be/db"
	"svetozar12/headless-cms-be/models"
	"svetozar12/headless-cms-be/routes/content/contentModel"
	field "svetozar12/headless-cms-be/routes/field"
	fieldtype "svetozar12/headless-cms-be/routes/fieldType"

	"github.com/gofiber/fiber/v2"
)

func ContentRoutes(app fiber.Router) {
	content := app.Group("/content")
	content.Get("/:id", getContentById)
	content.Get("/", getContent)
	content.Post("/", createContent)
	content.Put("/:id", updateContent)
	content.Delete("/:id", deleteContent)
}

// Content godoc
// @Summary      Get content by id
// @Tags         content
// @Accept       json
// @Param id     path int true "ID"
// @Success      200  {object} contentModel.Content
// @Router       /v1/content/{id} [get]
func getContentById(c *fiber.Ctx) error {
	var content contentModel.Content
	id := c.Params("id")

	db.DB.Preload("ContentModel").First(&content, id)
	return c.Status(fiber.StatusOK).JSON(content)
}

// Content godoc
// @Summary      Get all content
// @Tags         content
// @Accept       json
// @Param        page     query     int  false  "page"   default(1)
// @Param        limit    query     int  false  "limit"  default(10)
// @Param        userId  query     string  true   "userId"
// @Success      200  {object} models.PaginationModel[[]contentModel.Content]
// @Router       /v1/content [get]
func getContent(c *fiber.Ctx) error {
	var content []contentModel.Content
	var total int64
	page, _ := strconv.Atoi(c.Query("page"))
	limit, _ := strconv.Atoi(c.Query("limit"))
	userId := c.Query("userId")
	fmt.Println(userId)
	offSet := (page - 1) * limit
	db.DB.Where("user_id = ?", userId).Preload("ContentModel").Find(&content).Count(&total)
	db.DB.Where("user_id = ?", userId).Preload("ContentModel").Offset(offSet).Limit(limit).Find(&content)
	return c.Status(fiber.StatusOK).JSON(models.PaginationModel[[]contentModel.Content]{Pagination: models.Pagination{Total: total, Offset: page, Limit: limit}, Data: content})
}

// Content godoc
// @Summary      Create content
// @Tags         content
// @Accept       json
// @Param request body contentModel.Body true "query params""
// @Success      201  {object} contentModel.Content
// @Router       /v1/content [post]
func createContent(c *fiber.Ctx) error {
	content := new(contentModel.Content)
	var fieldTypes []fieldtype.FieldType
	err := c.BodyParser(content)
	if err != nil {
		return c.SendStatus(fiber.ErrUnprocessableEntity.Code)
	}

	db.DB.Where("content_model_id = ?", content.ModelId).Find(&fieldTypes)
	db.DB.Create(&content)
	db.DB.Preload("ContentModel").First(&content, content.ID)

	var fields []field.Field
	for i := 0; i < len(fieldTypes); i++ {
		fieldType := fieldTypes[i]
		field := field.Field{FieldType: fieldType, Body: field.Body{Name: fieldType.Name, Value: "", TypeId: int(fieldType.ID), ContentId: int(content.ID)}}
		fields = append(fields, field)
	}
	db.DB.Create(&fields)
	return c.Status(fiber.StatusCreated).JSON(content)
}

// Content godoc
// @Summary      Update content
// @Tags         content
// @Accept       json
// @Produce      json
// @Param request body contentModel.Body true "query params"
// @Param id     path int true "ID"
// @Success      200  {object}   contentModel.Content
// @Router       /v1/content/{id} [put]
func updateContent(c *fiber.Ctx) error {
	content := new(contentModel.Content)
	id := c.Params("id")
	err := c.BodyParser(content)
	if err != nil {
		return c.SendStatus(fiber.ErrUnprocessableEntity.Code)
	}
	db.DB.Where("id = ?", id).Updates(&content)
	return c.Status(fiber.StatusOK).JSON(content)
}

// Content godoc
// @Summary      Delete content
// @Tags         content
// @Accept       json
// @Param id     path int true "ID"
// @Success      200  {string} ok
// @Router       /v1/content/{id} [delete]
func deleteContent(c *fiber.Ctx) error {
	id := c.Params("id")
	var content contentModel.Content
	result := db.DB.Delete(&content, id)
	if result.RowsAffected == 0 {
		return c.SendStatus(fiber.StatusNotFound)
	}
	return c.SendStatus(fiber.StatusOK)
}
